docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+# Development Best Practices
+
+This document describes the best practices to follow when developing operators or other application components such as partitoners, stream codecs etc.
+
+## Operators
+
+These are general guidelines for all operators and the next sections talk about special considerations for input and output operators.
+
+* When writing a new operator to be used in an application, consider breaking it down into
+	* An abstract operator that encompasses the core functionality but leaves application specific schemas and logic to the implementation
+	* An optional concrete operator also in the library that extends the abstract operator and provides commonly used schema types such as strings, byte[] or POJOs.
+* Follow the conventions of the life cycle methods
+	* Do one time initialization of entities that apply for entire lifetime of the operator in the **setup** method, e.g., factory initializations. Initializations in **setup** are done in the container where the operator is deployed. Initializations in the constructor are discouraged as they lead to extra garbage in memory. This reason is as follows. The operator is instantiated on the client from where the application is being launched, then serialized and made available for launch. So the constructor is first called on the client and if it were to initialize any of the fields, that state would be saved during serialization. In the launch process the operator is deserialized. This would invoke the constructor again, which will initialize the fields but their state will get overwritten by the serialized state and the initial values would become garbage in memory.
```

Comment:
for entire => for the entire
This reason is as follows. => The reason is as follows:
is being launched => is launched
Initializations in the constructor are discouraged => Initializations of non-transient fields are discouraged
{seems like this also applies to initialization at the definition point ?}